### PR TITLE
Export details

### PR DESCRIPTION
**Pull request by wxsBSD**
These commits improve export parsing and exposes more information about exports in the PE module. Specifically it exposes:

pe.dll_name: The name of the dll from the export directory.
pe.export_timestamp: The timestamp found in the export directory.

It also exposes an array called pe.export_details[] which contains:

pe.export_details[].offset: The file offset where the exported function starts.
pe.export_details[].name: The name of the exported function.
pe.export_details[].forward_name: If the export is a forwarding function this is the name of the function it is forwarding to.
pe.export_details[].ordinal: The ordinal of the exported function after the ordinal base has been applied.

It is worth noting that prior to this change the ordinal recorded did NOT include the ordinal base, which meant it was not matching how most other tools operate. I confirmed that both pestudio and IDA display the ordinal with respect to the ordinal base when showing the exports. With this change we are now respecting the ordinal base which means any rules which did pe.exports(5) when the ordinal base is greater than 5 will no longer work (the actual exported ordinal is 10). I consider this a bug fix and am happy to break existing rules since I don't think there will be many.

My main goal was exposing the function offsets, but to expose them in a useful way I needed to create a pe.export_details array so I figured I may as well also expose the other fields with it. Eventually I'd love to see pe.exports() deprecated in favor of using the pe.export_details array.

To facilitate more interesting rules I added functions which will return the index into pe.export_details where certain things are true. For example, you can now write rules to find two exported functions with different names that point to the same offset with:

pe.export_details[pe.exports_index("a")].offset == pe.export_details[pe.exports_index("b")].offset

Since I'm storing all the useful data in pe.export_details now it made sense to delete the linked list used internally for storing export details and convert the existing pe.exports() to use the exposed data directly.

Lastly, I've tested this pretty extensively with some interesting DLLs and haven't found any problems but if anyone has any that they think might trip up my parsing I'm happy to take a look.

### CODE REVIEW

File: libyara/sizedstr.c [hunk 1]
```diff
@@ -27,11 +27,38 @@ ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
...
 
+int sized_string_cmp_nocase(
+  SIZED_STRING* s1,
+  SIZED_STRING* s2)
+{
+  size_t i = 0;
+
+  while (s1->length > i &&
+         s2->length > i &&
+         tolower(s1->c_string[i]) == tolower(s2->c_string[i]))
```
[the remark this answers is not in the document]
Comment (wxsBSD) [reply libyara/sizedstr.c]: Fixed, thanks for pointing that out!

**Comment by wxsBSD** [comment]
Added a few test cases and a legit Windows 10 DLL that exercises the code pretty well.

**Comment by edeca** [comment]
These features would be excellent to avoid writing rules like [this one](https://gist.github.com/edeca/cdc1657fa8a46b8ba45ad732377c035c).

**Comment by edeca** [comment]
A quick example of [before](https://gist.github.com/edeca/cdc1657fa8a46b8ba45ad732377c035c) and [after](https://gist.github.com/edeca/e8f05c457439a4f13e85f4b2dc68aa38) using this feature.

The proposed changes make a huge difference to rule readability, and offer some interesting options for hunting malicious files (primarily DLLs):

* Files exporting known functions where the "internal name" is not set correctly.  This is surprisingly common for DLL planting / side-loading where an attacker drops a known DLL (e.g. `libcef.dll`) but the internal name is set to something else (e.g. `test.dll`).
* Files exporting a named function with "junk" code, for example where the attacker implements one exported function and sets the rest to `xor eax,eax; retn`.

**Comment by wxsBSD** [comment]
Any chance this will make it in for 3.11.0? It provides some nice functionality for hunting for DLLs with stomped exports (often seen with sideloading attacks).

**Comment by plusvic** [comment]
Merged in d9f53012a54e4ccc1c81572277126a399d4fbb6a

**Pull request merged**